src/cryptography/hazmat/backends/openssl/dh.py

```diff
@@ -0,0 +1,205 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import, division, print_function
+
+from cryptography import utils
+from cryptography.hazmat.primitives.asymmetric import dh
+
+
+def _dh_cdata_to_parameters(dh_cdata, backend):
+    lib = backend._lib
+    ffi = backend._ffi
+
+    param_cdata = lib.DH_new()
+    backend.openssl_assert(param_cdata != ffi.NULL)
+    param_cdata = ffi.gc(param_cdata, lib.DH_free)
+
+    p = ffi.new("BIGNUM **")
+    g = ffi.new("BIGNUM **")
+    lib.DH_get0_pqg(dh_cdata, p, ffi.NULL, g)
+    backend.openssl_assert(p[0] != ffi.NULL)
+    backend.openssl_assert(g[0] != ffi.NULL)
+    p_dup = lib.BN_dup(p[0])
+    g_dup = lib.BN_dup(g[0])
+    backend.openssl_assert(p_dup != ffi.NULL)
+    backend.openssl_assert(g_dup != ffi.NULL)
+    res = lib.DH_set0_pqg(param_cdata, p_dup, ffi.NULL, g_dup)
+    backend.openssl_assert(res == 1)
+
+    return _DHParameters(backend, param_cdata)
+
+
+@utils.register_interface(dh.DHParametersWithSerialization)
+class _DHParameters(object):
+    def __init__(self, backend, dh_cdata):
+        self._backend = backend
+        self._dh_cdata = dh_cdata
+
+    def parameter_numbers(self):
+        p = self._backend._ffi.new("BIGNUM **")
+        g = self._backend._ffi.new("BIGNUM **")
+        self._backend._lib.DH_get0_pqg(self._dh_cdata,
+                                       p, self._backend._ffi.NULL, g)
+        self._backend.openssl_assert(p[0] != self._backend._ffi.NULL)
+        self._backend.openssl_assert(g[0] != self._backend._ffi.NULL)
+        return dh.DHParameterNumbers(
+            p=self._backend._bn_to_int(p[0]),
+            g=self._backend._bn_to_int(g[0])
+        )
+
+    def generate_private_key(self):
+        return self._backend.generate_dh_private_key(self)
+
+
+def _handle_dh_compute_key_error(errors, backend):
+    lib = backend._lib
+
+    backend.openssl_assert(errors[0][1:] == (
```

Comment:
There are 4 possible error values:
1. `DH_R_MODULUS_TOO_LARGE` - This error value is generated when p is larger then `OPENSSL_DH_MAX_MODULUS_BITS` that is defined to 10000. This error should not happen on any valid flow.
2. `DH_R_NO_PRIVATE_VALUE` - Which happens if the private key is NULL and should not happen in any valid flow.
3. `DH_R_INVALID_PUBKEY` - That we say might happen when the public key is invalid for this exhange.
4. `ERR_R_BN_LIB` - Happens when `bn_mod_exp` fail.

As you may see the only error value that is a user error is `DH_R_INVALID_PUBKEY`.